### ISSUE TITLE
purpose behind no-did-mount-set-state

### ISSUE THREAD

**Issue by jpollard-cs** [yannickcr/eslint-plugin-react#554]
What is the purpose behind the `no-did-mount-set-state` rule? I don't understand what's wrong with using `this.setState` inside of `componentDidMount`. Is this a legacy thing?

**Comment by yannickcr** [comment]
From [Alex Early's React Tips and Best Practices](http://aeflash.com/2015-02/react-tips-and-best-practices.html):

> Probably the worst thing you can do it call `setState` right after a render in `componentDidMount()` or `componentDidUpdate()`. While it is tempting to read things from the generated DOM at this point, it means most `render()` calls will actually call render twice. If a child component does the same thing, it could have up to four `render()` calls: two from its parent's render, and 2 from its own internal `setState()`. As you compose components this can lead to property/layout thrashing and exponentially more re-renders as your component hierarchy depth increases.

**Comment by ljharb** [comment]
Sadly, there's no cleaner way to run code in the browser that doesn't run when server-rendering (since `componentWillMount` runs on the server)

**Comment by jpollard-cs** [comment]
Valid points although I believe the following is an acceptable use-case that is of a nature that is fairly common:

```jsx
import React, { Component, PropTypes } from 'react';

class Dashboard extends Component {
    constructor(props, context) {
        super(props, context);
       
        this.state = { someElementStyle: { maxHeight: 200 } };
        this.setRef = ::this.setRef;
    }
    componentDidMount() {
        const node = ReactDOM.findDOMNode(this.someRef);
        const newHeight = window.innerHeight - node.offsetTop - 30;
        this.setState({ someElementStyle: { maxHeight: newHeight } });
    }

    render() {
        return (
            <div id="some-element-id" style={this.state.someElementStyle} ref={this.setRef}>
            </div>
        );
    }

    setRef(ref) {
        this.someRef= ref;
    }
}
```

AFIK there's not much of a better way of going about something like this - my concern is that folks will work around the error by doing something like `node.style.maxHeight = `${newHeight}px`;` which seems like it would be even worse as the Component will now be out of sync with the DOM. Thoughts?

**Comment by ljharb** [comment]
@jpollard-cs we just disable the rule in situations like this.

**Comment by jpollard-cs** [comment]
@ljharb yeah that's what I've done - just wanted to make sure there wasn't any other reason this was disallowed. Thanks!
